nighthawk/source/client/client.cc

```diff
@@ -0,0 +1,224 @@
+  } catch (const UriException) {
+    tls.shutdownGlobalThreading();
+    return false;
+  }
+
+  uint32_t concurrency = determineConcurrency();
+  std::vector<ClientWorkerPtr> workers;
+  // We try to offset the start of each thread so that workers will execute tasks evenly spaced in
+  // time.
+  // TODO(oschaaf): Expose the hard-coded two seconds below in configuration.
+  auto first_worker_start = time_system.monotonicTime() + 2s;
+  double inter_worker_delay_usec = (1. / options_->requests_per_second()) * 1000000 / concurrency;
+
+  for (uint32_t worker_number = 0; worker_number < concurrency; worker_number++) {
+    auto worker_delay = std::chrono::duration_cast<std::chrono::nanoseconds>(
+        ((inter_worker_delay_usec * worker_number) * 1us));
+    workers.push_back(std::make_unique<ClientWorkerImpl>(
+        api, tls, benchmark_client_factory, sequencer_factory, uri, store_factory.create(),
+        worker_number, first_worker_start + worker_delay));
+  }
+
+  Envoy::Runtime::RandomGeneratorImpl generator;
+  Envoy::Runtime::ScopedLoaderSingleton loader(
+      Envoy::Runtime::LoaderPtr{new Envoy::Runtime::LoaderImpl(generator, *store, tls)});
+
+  for (auto& w : workers) {
+    w->start();
+  }
+
+  bool ok = true;
+  for (auto& w : workers) {
+    w->waitForCompletion();
+    ok = ok && w->success();
+  }
+  if (ok) {
+    merged_statistics = mergeWorkerStatistics(statistic_factory, workers);
+    merged_counters = mergeWorkerCounters(workers);
+  }
+
+  tls.shutdownGlobalThreading();
+  return ok;
+}
+
+bool Main::run() {
+  Envoy::Thread::MutexBasicLockable log_lock;
+  std::vector<StatisticPtr> merged_statistics;
+  std::map<std::string, uint64_t> merged_counters;
+  auto logging_context = std::make_unique<Envoy::Logger::Context>(
+      spdlog::level::from_str(options_->verbosity()), "[%T.%f][%t][%L] %v", log_lock);
+  BenchmarkClientFactoryImpl benchmark_client_factory(*options_);
+  SequencerFactoryImpl sequencer_factory(*options_);
+
+  std::cout << "Nighthawk - A layer 7 protocol benchmarking tool.\n";
+
+  if (runWorkers(benchmark_client_factory, sequencer_factory, merged_statistics, merged_counters)) {
+    Envoy::RealTimeSource time_source;
+    ConsoleOutputFormatterImpl console_formatter(time_source, *options_, merged_statistics,
+                                                 merged_counters);
+    // TODO(oschaaf): output format, location, method, etc should be optionized.
+    std::cout << console_formatter.toString();
+    JsonOutputFormatterImpl json_formatter(time_source, *options_, merged_statistics,
+                                           merged_counters);
+    mkdir("measurements", 0777);
+    std::ofstream stream;
+    const int64_t epoch_seconds = time_source.systemTime().time_since_epoch().count();
+    std::string filename = fmt::format("measurements/{}.json", epoch_seconds);
+    stream.open(filename);
+    stream << json_formatter.toString();
```

Comment:
This and `runWorkers` are pretty big functions, is there a reasonable way to factor out bits, e.g. output formatting here?